integrations/Matomo/util.js

```diff
@@ -0,0 +1,291 @@
+        break;
+      }
+
+      price = get(message, "properties.price");
+      if (!price) {
+        logger.error("User parameter (sku or product_id) is required");
+        break;
+      }
+      window._paq.push([
+        "setEcommerceView",
+        productSKU,
+        productName,
+        categoryName,
+        price,
+      ]);
+      window._paq.push(["trackPageView"]);
+      break;
+
+    case "ADD_ECOMMERCE_ITEM":
+      productSKU =
+        get(message, "properties.sku") || get(message, "properties.product_id");
+      if (!productSKU) {
+        logger.error("User parameter (sku or product_id) is required");
+        break;
+      }
+
+      productName = get(message, "properties.name");
+      categoryName = get(message, "properties.category");
+      price = get(message, "properties.price");
+      quantity = get(message, "properties.quantity");
+      window._paq.push([
+        "addEcommerceItem",
+        productSKU,
+        productName,
+        categoryName,
+        price,
+        quantity,
+      ]);
+      break;
+
+    case "REMOVE_ECOMMERCE_ITEM":
+      productSKU =
+        get(message, "properties.sku") || get(message, "properties.product_id");
+      if (!productSKU) {
+        logger.error("User parameter (sku or product_id) is required");
+        break;
+      }
+      window._paq.push(["removeEcommerceItem", productSKU]);
+      break;
+
+    case "TRACK_ECOMMERCE_ORDER":
+      orderId = get(message, "properties.order_id");
+      if (!orderId) {
+        logger.error("User parameter order_id is required");
+        break;
+      }
+
+      grandTotal =
+        get(message, "properties.total") || get(message, "properties.revenue");
+      /** grandTotal is a mandatory field
+       * If grandTotal is not found inside the properties, we are calculating it manually
+       */
+      if (!grandTotal) {
+        if (products && products.length > 0) {
+          grandTotal = 0;
+          // iterating through the products array and calculating grandTotal
+          for (let product = 0; product < products.length; product++) {
+            if (product.price && product.quantity) {
+              grandTotal += product.price * product.quantity;
+            }
+          }
+        }
+      }
+
+      /** subTotal is not a listed property in "Order Completed" event
+       * if user doesn't provide this property, then we are calculating it from grandTotal
+       * ref: https://matomo.org/faq/reports/analyse-ecommerce-reporting-to-improve-your-sales/#conversions-overview
+       */
+      subTotal = get(message, "properties.subTotal");
+      if (!subTotal) {
+        subTotal = grandTotal;
+      }
+      tax = get(message, "properties.tax");
+      shipping = get(message, "properties.shipping");
+      discount = get(message, "properties.discount");
+
+      // ref: https://matomo.org/faq/reports/analyse-ecommerce-reporting-to-improve-your-sales/#conversions-overview
+      grandTotal += shipping + tax;
```

Comment:
Check if shipping and tax are present or not and are numbers or not.

Reply:
Done